numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -2142,6 +2142,238 @@ string_inputs_promoter(
+        npy_static_string os = {0, NULL};
+        npy_packed_static_string *ops = (npy_packed_static_string *)optr;
+        int is_isnull = NpyString_load(iallocator, ips, &is);
+        if (is_isnull == -1) {
+            npy_gil_error(PyExc_MemoryError, "Failed to load string in slice");
+            goto fail;
+        }
+        else if (is_isnull) {
+            npy_gil_error(PyExc_TypeError, "Cannot slice null string");
+            goto fail;
+        }
+
+        // adjust slice to string length in codepoints
+        // and handle negative indices
+        Buffer<ENCODING::UTF8> inbuf((char *)is.buf, is.size);
+        size_t num_codepoints = inbuf.num_codepoints();
+        npy_intp slice_length =
+                PySlice_AdjustIndices(num_codepoints, &start, &stop, step);
+
+        // move to start position in inbuf
+        inbuf += start;
+
+        // first pass: compute outsize
+        size_t outsize = 0;
+        if (step > 0) {
+            const char *inbuf_ptr = inbuf.buf;
+            npy_intp i_idx = 0, o_idx = 0;
+            while (o_idx < slice_length) {
+                int num_bytes = num_bytes_for_utf8_character(
+                        (const unsigned char *)inbuf_ptr);
+
+                if (i_idx % step == 0) {
+                    outsize += num_bytes;
+                    o_idx++;
+                }
+
+                inbuf_ptr += num_bytes;
+                i_idx++;
+            }
+        }
+        else {
+            if (slice_length > 0) {
+                const char *inbuf_ptr = inbuf.buf;
+
+                // get number of bytes for current character
+                int num_bytes = num_bytes_for_utf8_character(
+                        (const unsigned char *)inbuf_ptr);
+
+                npy_intp i_idx = 0, o_idx = 0;
+                while (o_idx < slice_length) {
+                    if (i_idx % (-step) == 0) {
+                        outsize += num_bytes;
+                        o_idx++;
+                    }
+
+                    // break out of loop if we would read past the beginning of
+                    // the buffer
+                    if (inbuf_ptr <= is.buf) {
+                        break;
+                    }
+
+                    char *previous_char_ptr =
+                            (char *)find_previous_utf8_character(
+                                    (const unsigned char *)inbuf_ptr, 1);
+                    num_bytes = inbuf_ptr - previous_char_ptr;
+                    inbuf_ptr = previous_char_ptr;
+                    i_idx++;
+                }
+            }
+        }
+
+        char *buf = NULL;
+        if (context->descriptors[0] == context->descriptors[4]) {
+            // in-place
+            buf = (char *)PyMem_RawMalloc(outsize);
+            if (buf == NULL) {
+                npy_gil_error(PyExc_MemoryError,
+                              "Failed to allocate string in slice");
+                goto fail;
+            }
+        }
```

Comment:
we're not allowing in-place operations yet so you can just delete the case above. In principle if we add it in the future we'd need to redo this but better to do that than have dead code IMO.

I know there are other places we have dead code you're copying from, but was probably a mistake on my part to allow that, better to not add new dead code.